File: src/ast.py
from abc import ABC, abstractmethod
from src.tokens import Token

class ASTNode(ABC):
    @abstractmethod
    def __str__(self) -> str:
        pass

class Statement(ASTNode, ABC):
    """Base class for all action-oriented structures (let, const, return)."""
    pass

class Expression(ASTNode, ABC):
    """Base class for all value-producing structures (literals, math, names)."""
    pass

class Program(ASTNode):
    """The root node of the entire source code file."""
    def __init__(self):
        self.statements: list[Statement] = []

    def __str__(self) -> str:
        return "\n".join(str(stmt) for stmt in self.statements)

# --- Statements ---

class LetStatement(Statement):
    """Represents 'let x = <expression>' or 'let x: Type = <expression>'"""
    def __init__(self, token: Token, name: 'Identifier', explicit_type: 'Identifier' = None, value: Expression = None):
        self.token = token               # The TokenType.LET token
        self.name = name                 # Identifier node for the variable name
        self.explicit_type = explicit_type # Optional explicit type Identifier node
        self.value = value               # Expression node evaluated to assign

    def __str__(self) -> str:
        type_str = f": {self.explicit_type}" if self.explicit_type else ""
        return f"let {self.name}{type_str} = {self.value}"

class ConstStatement(Statement):
    """Represents 'const x = <expression>'"""
    def __init__(self, token: Token, name: 'Identifier', explicit_type: 'Identifier' = None, value: Expression = None):
        self.token = token               # The TokenType.CONST token
        self.name = name                 # Identifier node
        self.explicit_type = explicit_type 
        self.value = value

    def __str__(self) -> str:
        type_str = f": {self.explicit_type}" if self.explicit_type else ""
        return f"const {self.name}{type_str} = {self.value}"

# --- Expressions ---

class Identifier(Expression):
    """Represents a variable reference or type name (e.g., 'age', 'Int')."""
    def __init__(self, token: Token, value: str):
        self.token = token
        self.value = value               # The actual string literal name

    def __str__(self) -> str:
        return self.value

class IntegerLiteral(Expression):
    """Represents an integer value node (e.g., 42)."""
    def __init__(self, token: Token, value: int):
        self.token = token
        self.value = value

    def __str__(self) -> str:
        return str(self.value)

class FloatLiteral(Expression):
    """Represents a float value node (e.g., 3.14)."""
    def __init__(self, token: Token, value: float):
        self.token = token
        self.value = value

    def __str__(self) -> str:
        return str(self.value)

class StringLiteral(Expression):
    """Represents a string value node (e.g., "Hello Aura")."""
    def __init__(self, token: Token, value: str):
        self.token = token
        self.value = value

    def __str__(self) -> str:
        return f'"{self.value}"'

File: src/ast.rs
use crate::tokens::Token;

// The base trait that all nodes in the Aura tree must support
pub trait Node {
    fn to_string(&self) -> String;
}

// --- Statement Types ---
#[allow(dead_code)]
#[derive(Debug, Clone, PartialEq)]
pub enum Statement {
    Let(LetStatement),
    Const(ConstStatement),
    Expression(Expression), // Allows standalone values or function calls to act as statements
}

impl Node for Statement {
    fn to_string(&self) -> String {
        match self {
            Statement::Let(stmt) => stmt.to_string(),
            Statement::Const(stmt) => stmt.to_string(),
            Statement::Expression(expr) => expr.to_string(),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct LetStatement {
    pub token: Token,              // The Token::Let token
    pub name: String,              // The variable identifier name
    pub explicit_type: Option<String>, // Optional explicit type annotation (e.g., Some("Int"))
    pub value: Expression,         // The inner expression value being assigned
}

impl Node for LetStatement {
    fn to_string(&self) -> String {
        let type_str = match &self.explicit_type {
            Some(t) => format!(": {}", t),
            None => "".to_string(),
        };
        format!("let {}{} = {}", self.name, type_str, self.value.to_string())
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct ConstStatement {
    pub token: Token,              // The Token::Const token
    pub name: String,              // The constant identifier name
    pub explicit_type: Option<String>,
    pub value: Expression,
}

impl Node for ConstStatement {
    fn to_string(&self) -> String {
        let type_str = match &self.explicit_type {
            Some(t) => format!(": {}", t),
            None => "".to_string(),
        };
        format!("const {}{} = {}", self.name, type_str, self.value.to_string())
    }
}

// --- Expression Types ---

#[derive(Debug, Clone, PartialEq)]
pub enum Expression {
    Identifier(String),
    IntegerLiteral(i64),
    FloatLiteral(f64),
    StringLiteral(String),
}

impl Node for Expression {
    fn to_string(&self) -> String {
        match self {
            Expression::Identifier(val) => val.clone(),
            Expression::IntegerLiteral(val) => val.to_string(),
            Expression::FloatLiteral(val) => val.to_string(),
            Expression::StringLiteral(val) => format!("\"{}\"", val),
        }
    }
}

// --- Program Root Node ---

pub struct Program {
    pub statements: Vec<Statement>,
}

impl Node for Program {
    fn to_string(&self) -> String {
        self.statements
            .iter()
            .map(|stmt| stmt.to_string())
            .collect::<Vec<String>>()
            .join("\n")
    }
}
